projects/pgai/pgai/vectorizer/migrations/__init__.py

```diff
@@ -0,0 +1,131 @@
+import importlib
+import pkgutil
+import sys
+from collections.abc import Callable
+from dataclasses import dataclass
+from typing import Any, Generic, TypeVar
+
+import semver
+import structlog
+
+logger = structlog.get_logger()
+
+# config generic type
+C = TypeVar("C")
+MigrationFunc = Callable[[C], dict[str, Any]]
+
+
+@dataclass
+class Migration(Generic[C]):
+    version: str  # target version
+    source_config_class: type[C]  # config dataclass for the source version
+    apply: MigrationFunc[C]
+    description: str = ""
+
+
+# registry to hold all migrations
+migrations: list[Migration[Any]] = []
+
+MigrationDecorator = Callable[[MigrationFunc[C]], MigrationFunc[C]]
+
+
+def register_migration(
+    version: str, source_config_class: type[C], description: str = ""
+) -> MigrationDecorator[C]:
+    """
+    Decorator to register a migration function with
+    config class validation
+    """
+
+    def decorator(func: MigrationFunc[C]) -> MigrationFunc[C]:
+        migrations.append(
+            Migration(
+                version=version,
+                source_config_class=source_config_class,
+                apply=func,
+                description=description,
+            )
+        )
+        migrations.sort(key=lambda m: semver.VersionInfo.parse(m.version))
+        return func
+
+    return decorator
+
+
+def get_latest_version() -> str:
+    """Get the latest migration version available"""
+    if not migrations:
+        return "0.0.0"
+    return migrations[-1].version
+
+
+def apply_migrations(data: dict[str, Any]) -> dict[str, Any]:
+    """
+    Apply all necessary migrations incrementally until reaching the latest version
+
+    Args:
+        data: The config data to migrate
+
+    Returns:
+        The migrated config data
+    """
+    if not data:
+        return data
+
+    # determine starting version
+    current_version = data.get("version")
+    if current_version is None:
+        logger.warning(
+            "Vectorizer Config migration can't be considered. "
+            "Raw data does not contain a version field."
```

Comment:
I don't really like "can't be considered" here - I don't really understand what it's trying to say. How about:
```suggestion
            "Unable to migrate configuration: 'version' field missing."
```